go/src/koding/klient/machine/client/cached.go

```diff
@@ -0,0 +1,212 @@
+		mu.Lock()
+		aUsername, aKeys = username, keys
+		rErr = err
+		lastCall = time.Now()
+		mu.Unlock()
+
+		return
+	}
+}
+
+// MountHeadIndex calls registered Client's MountHeadIndex method and caches its
+// result for the specified interval. It doesn't cache results from disconnected
+// client. If call arguments change, the cache will be invalidated.
+func (c *Cached) MountHeadIndex(path string) (string, int, int64, error) {
+	return c.mhi(path)
+}
+
+func mhi(c Client, interval time.Duration) func(string) (string, int, int64, error) {
+	lastCall, mu := time.Now().Add(-interval-time.Second), sync.Mutex{}
+
+	aPath := ""
+	rAbsPath, rCount, rDiskSize, rErr := "", 0, int64(0), error(nil)
+
+	return func(path string) (absPath string, count int, diskSize int64, err error) {
+		mu.Lock()
+		if time.Since(lastCall) < interval && aPath == path && rErr != ErrDisconnected {
+			absPath, count, diskSize, err = rAbsPath, rCount, rDiskSize, rErr
+			mu.Unlock()
+			return
+		}
+		mu.Unlock()
+
+		absPath, count, diskSize, err = c.MountHeadIndex(path)
+
+		mu.Lock()
+		aPath = path
+		rAbsPath, rCount, rDiskSize, rErr = absPath, count, diskSize, err
+		lastCall = time.Now()
+		mu.Unlock()
+
+		return
+	}
+}
+
+// MountGetIndex calls registered Client's MountGetIndex method and caches its
+// result for the specified interval. It doesn't cache results from disconnected
+// client. If call arguments change, the cache will be invalidated.
+func (c *Cached) MountGetIndex(path string) (*index.Index, error) {
+	return c.mgi(path)
+}
+
+func mgi(c Client, interval time.Duration) func(string) (*index.Index, error) {
+	lastCall, mu := time.Now().Add(-interval-time.Second), sync.Mutex{}
+
+	aPath := ""
+	rIdx, rErr := (*index.Index)(nil), error(nil)
+
+	return func(path string) (idx *index.Index, err error) {
+		mu.Lock()
+		if time.Since(lastCall) < interval && aPath == path && rErr != ErrDisconnected {
+			idx, err = rIdx, rErr
+			mu.Unlock()
+			return
+		}
+		mu.Unlock()
+
+		idx, err = c.MountGetIndex(path)
+
+		mu.Lock()
+		aPath = path
+		rIdx, rErr = idx, err
+		lastCall = time.Now()
+		mu.Unlock()
+
+		return
+	}
+}
+
+// DiskBlocks calls registered Client's DiskBlocks method and caches its result
+// for the specified interval. It doesn't cache results from disconnected
+// client. If call arguments change, the cache will be invalidated.
+func (c *Cached) DiskBlocks(path string) (uint64, uint64, uint64, uint64, error) {
+	return c.db(path)
+}
+
+func db(c Client, interval time.Duration) func(string) (uint64, uint64, uint64, uint64, error) {
+	lastCall, mu := time.Now().Add(-interval-time.Second), sync.Mutex{}
+
+	aPath := ""
+	rSize, rTotal, rFree, rUsed, rErr := uint64(0), uint64(0), uint64(0), uint64(0), error(nil)
+
+	return func(path string) (size, total, free, used uint64, err error) {
+		mu.Lock()
+		if time.Since(lastCall) < interval && aPath == path && rErr != ErrDisconnected {
+			size, total, free, used, err = rSize, rTotal, rFree, rUsed, rErr
```

Comment:
yes go provides a way to assign variables in one line but this is not readable after 3 vars :)